### PR TITLE
Implement MS.AAD.3.1v1 phishing resistant mfa for all users

### LINKED ISSUE #351: Update ScubaGear AAD 2.4 2.4 MFA policies per feedback and discussion
# 💡 Summary #

Update the ScubaGear Rego policy checks related to 2.4 MFA per issue #329.

## Acceptance criteria ##

How do we know when this work is done?

- [ ] Rego code and test cases are updated and unit tested
- [ ] Updates are tested in all test tenants

### PR DESCRIPTION
**Pull request by crutchfield**
#  #

## 🗣 Description ##

Verify a CAP exists for acceptable MFA Strengths applied to All users/app with only explicilty specified exclusions 

closes #351  - Will not auto close since [not default branch](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)?

## 💭 Motivation and context ##

Initial implementation of new policy

## 🧪 Testing ##

Added new unit test for policy. All pass.
Tested against tenants and got expected results. 
NOTE: Use caution if attempting to set conditions to pass policy in tenant because you can lock out users. 

## 📷 Screenshots (if appropriate) ##

image

image

## ✅ Pre-approval checklist ##

- [X] This PR has an informative and human-readable title.
- [X] Changes are limited to a single goal - *eschew scope creep!*
- [X] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [X] All relevant type-of-change labels have been added.
- [X] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [X] Tests have been added and/or modified to cover the changes in this PR.
- [X] All new and existing [applicable] tests pass. 

## ✅ Pre-merge checklist ##

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

## ✅ Post-merge checklist ##

- [ ] Create a release.

### CODE REVIEW

File: Rego/Utils/TestUtils.rego [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package test.utils
+import future.keywords
+
+#IsEmpty(null, _) = true
+#IsEmpty(a, b) if not a[b]
+#IsEmpty(a, b) if a[b] == ""
```
[the remark this answers is not in the document]
Comment (crutchfield) [reply Rego/Utils/TestUtils.rego]: Concur - removed.

File: Rego/AADConfig.rego [hunk 2]
```diff
@@ -3,6 +3,9 @@ import future.keywords
 import data.report.utils.NotCheckedDetails
 import data.report.utils.Format
 import data.report.utils.ReportDetailsBoolean
+import data.test.utils.IsEmptyContainer
+import data.test.utils.Contains
+import data.test.utils.Count
```
[the remark this answers is not in the document]
Comment (crutchfield) [reply Rego/AADConfig.rego]: Concur - renamed

**Review by ssatyapal123 — APPROVED** [review]
- Ran against G3, E5, and G5 tenants, all with expected results.
- Policy 3.1 fails due to prescribed testing protocol to refrain from requiring Phishing Resistant MFA for all users
- Rego unit tests all passed.
- G5 tenant: Added an allowed exclusion for the "EmergencyExempt" group from policy 3.2 to confirm the policy passed.

image


image

**Comment by nanda-katikaneni** [comment]
@crutchfield seems the branch has conflicts, please resolve them for the merger.

**Comment by crutchfield** [comment]
@nanda-katikaneni Conflicts resolved.

**Pull request merged**